panic: runtime error: invalid memory address or nil pointer dereference

I've been receiving the following error running routedns on a public server with DOH exposed.

Ubuntu 20.04.3 LTS (GNU/Linux 5.4.0-90-generic x86_64)
go version go1.17.3 linux/amd64

```
Dec 20 16:40:25 server routedns[203538]: time="2021-12-20T16:40:25Z" level=info msg="starting listener" addr=":443" id=local-doh protocol=doh
Dec 20 16:40:25 server routedns[203538]: time="2021-12-20T16:40:25Z" level=info msg="starting listener" addr=":53" id=local-udp protocol=udp
Dec 20 16:40:25 server routedns[203538]: time="2021-12-20T16:40:25Z" level=info msg="starting listener" addr=":53" id=local-tcp protocol=tcp
Dec 20 17:06:44 server routedns[203538]: panic: runtime error: invalid memory address or nil pointer dereference
Dec 20 17:06:44 server routedns[203538]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x817afe]
Dec 20 17:06:44 server routedns[203538]: goroutine 1993 [running]:
Dec 20 17:06:44 server routedns[203538]: github.com/miekg/dns.(*OPT).copy(0xc003b58980)
Dec 20 17:06:44 server routedns[203538]:         /opt/routedns/pkg/mod/github.com/miekg/dns@v1.1.43/ztypes.go:866 +0xfe
Dec 20 17:06:44 server routedns[203538]: github.com/miekg/dns.(*Msg).CopyTo(0xc00577cbd0, 0xc00577cc60)
Dec 20 17:06:44 server routedns[203538]:         /opt/routedns/pkg/mod/github.com/miekg/dns@v1.1.43/msg.go:1076 +0x277
Dec 20 17:06:44 server routedns[203538]: github.com/miekg/dns.(*Msg).Copy(...)
Dec 20 17:06:44 server routedns[203538]:         /opt/routedns/pkg/mod/github.com/miekg/dns@v1.1.43/msg.go:1050
Dec 20 17:06:44 server routedns[203538]: github.com/folbricht/routedns.(*Cache).Resolve(0xc0000b3810, 0xc00017b5f0, {{0xc0037f15cc, 0x8, 0x0}})
Dec 20 17:06:44 server routedns[203538]:         /opt/routedns/pkg/mod/github.com/folbricht/routedns@v0.1.0/cache.go:133 +0x434
Dec 20 17:06:44 server routedns[203538]: github.com/folbricht/routedns.listenHandler.func1({0xb59158, 0xc000157180}, 0xc00017b5f0)
Dec 20 17:06:44 server routedns[203538]:         /opt/routedns/pkg/mod/github.com/folbricht/routedns@v0.1.0/dnslistener.go:71 +0x60b
Dec 20 17:06:44 server routedns[203538]: github.com/miekg/dns.HandlerFunc.ServeDNS(0xc001757500, {0xb59158, 0xc000157180}, 0xc00017b5f0)
Dec 20 17:06:44 server routedns[203538]:         /opt/routedns/pkg/mod/github.com/miekg/dns@v1.1.43/server.go:37 +0x2f
Dec 20 17:06:44 server routedns[203538]: github.com/miekg/dns.(*Server).serveDNS(0xc0000d05a0, {0xc001757500, 0x80, 0xc000090298}, 0xc000157180)
Dec 20 17:06:44 server routedns[203538]:         /opt/routedns/pkg/mod/github.com/miekg/dns@v1.1.43/server.go:651 +0x4da
Dec 20 17:06:44 server routedns[203538]: github.com/miekg/dns.(*Server).serveTCPConn(0xc0000d05a0, 0x0, {0xb592b8, 0xc0000b6728})
Dec 20 17:06:44 server routedns[203538]:         /opt/routedns/pkg/mod/github.com/miekg/dns@v1.1.43/server.go:559 +0x26c
Dec 20 17:06:44 server routedns[203538]: created by github.com/miekg/dns.(*Server).serveTCP
Dec 20 17:06:44 server routedns[203538]:         /opt/routedns/pkg/mod/github.com/miekg/dns@v1.1.43/server.go:460 +0x2d9
```

Is this the same issue?
```go
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x8e9e46]

goroutine 16728 [running]:
github.com/folbricht/routedns.stripPadding(0xc01391c6c0)
	/root/go/src/github.com/folbricht/routedns/padding.go:98 +0xe6
github.com/folbricht/routedns.(*DNSClient).Resolve(0xc000469ec0, 0x7f4b3b8db008, {{0xc019c10c70, 0x406d9d, 0xc000352001}})
	/root/go/src/github.com/folbricht/routedns/dnsclient.go:73 +0x1db
github.com/folbricht/routedns.(*Fastest).Resolve.func1()
	/root/go/src/github.com/folbricht/routedns/fastest.go:42 +0x49
created by github.com/folbricht/routedns.(*Fastest).Resolve
	/root/go/src/github.com/folbricht/routedns/fastest.go:41 +0x665
```

--- a/dnsclient.go
+++ b/dnsclient.go
@@ -62,6 +62,9 @@ func NewDNSClient(id, endpoint, network string, opt DNSClientOptions) (*DNSClien
 
 // Resolve a DNS query.
 func (d *DNSClient) Resolve(q *dns.Msg, ci ClientInfo) (*dns.Msg, error) {
+	// Packing a message is not always a read-only operation, make a copy
+	q = q.Copy()
+
 	logger(d.id, q, ci).WithFields(logrus.Fields{
 		"resolver": d.endpoint,
 		"protocol": d.net,

--- a/dohclient.go
+++ b/dohclient.go
@@ -352,6 +352,9 @@ func (s *quicConnection) NextConnection() quic.Connection {
 }
 
 func quicRestart(s *quicConnection) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
 	// Try to open a new connection, but clean up our mess before we do so
 	// This function should be called with the quicConnection locked, but lock checking isn't provided
 	// in golang; the issue was closed with "Won't fix"

--- a/dohclient.go
+++ b/dohclient.go
@@ -96,6 +96,9 @@ func NewDoHClient(id, endpoint string, opt DoHClientOptions) (*DoHClient, error)
 
 // Resolve a DNS query.
 func (d *DoHClient) Resolve(q *dns.Msg, ci ClientInfo) (*dns.Msg, error) {
+	// Packing a message is not always a read-only operation, make a copy
+	q = q.Copy()
+
 	logger(d.id, q, ci).WithFields(logrus.Fields{
 		"resolver": d.endpoint,
 		"protocol": "doh",

--- a/dohclient.go
+++ b/dohclient.go
@@ -352,9 +352,6 @@ func (s *quicConnection) NextConnection() quic.Connection {
 }
 
 func quicRestart(s *quicConnection) error {
-	s.mu.Lock()
-	defer s.mu.Unlock()
-
 	// Try to open a new connection, but clean up our mess before we do so
 	// This function should be called with the quicConnection locked, but lock checking isn't provided
 	// in golang; the issue was closed with "Won't fix"

--- a/dotclient.go
+++ b/dotclient.go
@@ -68,6 +68,9 @@ func NewDoTClient(id, endpoint string, opt DoTClientOptions) (*DoTClient, error)
 
 // Resolve a DNS query.
 func (d *DoTClient) Resolve(q *dns.Msg, ci ClientInfo) (*dns.Msg, error) {
+	// Packing a message is not always a read-only operation, make a copy
+	q = q.Copy()
+
 	logger(d.id, q, ci).WithFields(logrus.Fields{
 		"resolver": d.endpoint,
 		"protocol": "dot",

--- a/dtlsclient.go
+++ b/dtlsclient.go
@@ -90,6 +90,9 @@ func NewDTLSClient(id, endpoint string, opt DTLSClientOptions) (*DTLSClient, err
 
 // Resolve a DNS query.
 func (d *DTLSClient) Resolve(q *dns.Msg, ci ClientInfo) (*dns.Msg, error) {
+	// Packing a message is not always a read-only operation, make a copy
+	q = q.Copy()
+
 	logger(d.id, q, ci).WithFields(logrus.Fields{
 		"resolver": d.endpoint,
 		"protocol": "dtls",

